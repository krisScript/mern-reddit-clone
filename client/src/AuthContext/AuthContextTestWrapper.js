import {AuthContextData} from './AuthContext'
import {BrowserRouter} from 'react-router-dom'
import React from 'react'
const AuthContextTestWrapper = ({authState,children,updateUserDataReducer}) => {
    return (
       <AuthContextData.Provider value={{authState,updateUserDataReducer}}>
        <BrowserRouter>
          {children}
          </BrowserRouter>
        </AuthContextData.Provider>
    )
}
export default AuthContextTestWrapper 