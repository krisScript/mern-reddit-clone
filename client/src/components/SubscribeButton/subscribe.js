const subscribe = async (communityId,token) => {
    const response = await fetch(`http://localhost:8080/subscribe/${communityId}`, {
      method: 'POST',
      headers: {
        Authorization: 'Bearer ' + token,
        'Content-Type': 'application/json'
      },
    });

    const responseData = await response.json();
    if (responseData.message === 'Validation failed.') {
      return false
    } else if (response.status  === 201) {
      return responseData.userData
    }
  };
export default subscribe