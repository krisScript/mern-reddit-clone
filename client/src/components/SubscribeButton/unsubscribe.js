const unsubscribe = async (communityId,token) => {
    const response = await fetch(`http://localhost:8080/unsubscribe/${communityId}`, {
      method: 'POST',
      headers: {
        Authorization: 'Bearer ' + token,
        'Content-Type': 'application/json'
      },
    });

    const responseData = await response.json();
    if (responseData.message === 'Validation failed.') {
      return false
    } else if (response.status  === 201) {
      return responseData.userData
    }
  };
export default unsubscribe