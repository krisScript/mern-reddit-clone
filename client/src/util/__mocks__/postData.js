const postData = () => {
    return Promise.resolve(
      {
       userData:{
         username:'updated user'
       }
       }
    );
  };
  export default postData;