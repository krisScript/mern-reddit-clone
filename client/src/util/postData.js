const postData = async (url,data,token) => {
  console.log('nani')
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: 'Bearer ' + token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const responseData = await response.json();
  if (response.status === 422) {
    const validationErrors = responseData.data
    return {
      validationErrors,
    };
  }
  return responseData
};
export default postData





