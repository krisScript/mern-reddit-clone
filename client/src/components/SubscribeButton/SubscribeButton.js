import React from 'react';
import { withRouter } from 'react-router-dom';
import postData from '../../util/postData';
import useAuthContext from '../../hooks/useAuthContext/useAuthContext'
const SubscribeButtop = ({history,id}) => {
  const { isAuth, token,userId, userData,updateUserDataReducer } = useAuthContext();

  let subscribed = false;
  if (isAuth) {
    subscribed = userData.communities.includes(id);
  }
  const message = subscribed ? 'Unsubscribe' : 'Subscribe';
  const subscribeHandler = async () => {
    if (isAuth) {
      const apiUrl = subscribed
        ? `http://localhost:8080/user/unsubscribe/${id}`
        : `http://localhost:8080/user/subscribe/${id}`;
      const responseData = await postData(apiUrl, {}, token);
      if (responseData.userData) {
        const authState = {isAuth,token,userId}
        updateUserDataReducer({
          authState,
          newUserData: responseData.userData
        });
      }
    } else history.push('/login');
  };
  return (
    <>
      <button onClick={subscribeHandler} className="button">
        {message}
      </button>
    </>
  );
};
export default withRouter(SubscribeButtop);
