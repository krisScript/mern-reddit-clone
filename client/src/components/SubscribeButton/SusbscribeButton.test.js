import React from 'react';

import { render, cleanup, fireEvent } from 'react-testing-library';
import AuthContextTestWrapper from '../../AuthContext/AuthContextTestWrapper'
import SubscribeButton from './SubscribeButton';
import postData from '../../util/postData';
jest.mock('../../util/postData');

  const id='1'
  const updateUserDataReducer = jest.fn()
   
describe('<SubscribeButton />', () => {
    describe('testing subscribe button',() => {
        const authState =  {isAuth:true,token:'sometoken',userId:'someUserId',userData:{communities:[]}}
        const { container, rerender,getByTestId,getByText} = render (<SubscribeButton id={id} />,{wrapper: ({ children }) => (
        <AuthContextTestWrapper authState={authState} updateUserDataReducer={updateUserDataReducer} children={children} />
      )})

      beforeEach(() => {
        const subButton = getByText('Subscribe')
        fireEvent.click(subButton)
    })
      it('snapshot', () => {
        expect(container).toMatchSnapshot();
      });
      it('button with text Subscribe text should exist',() => {
        const subButton = getByText('Subscribe')
        expect(subButton).toBeTruthy()
      })
     
      it('should render with the correct name,placeholder,value,className and type', async () => {
         expect(updateUserDataReducer).toBeCalled()
        });

        
    })
  describe('user is subscribed',()=> {
    const newAuthState = {isAuth:true,token:'sometoken',userId:'someUserId',userData:{communities:['1']}}
    const { container, rerender,getByTestId,getByText} = render(<SubscribeButton id={id} />,{wrapper: ({ children }) => (
       <AuthContextTestWrapper authState={newAuthState} updateUserDataReducer={updateUserDataReducer} children={children} />
     )})
    it('button with Unsubscribe test should exist', () => {
          const subButton = getByText('Unsubscribe')
          expect(subButton).toBeTruthy()
    })
  })
 
});
