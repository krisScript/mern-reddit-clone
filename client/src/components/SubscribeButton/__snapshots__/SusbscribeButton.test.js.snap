// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscribeButton /> testing subscribe button snapshot 1`] = `
<div>
  <button
    class="button"
  >
    Subscribe
  </button>
</div>
`;
